pkg/substitution/substitution.go

```diff
@@ -58,6 +67,15 @@ func ValidateVariableP(value, prefix string, vars sets.String) *apis.FieldError
 				}
 			}
 		}
+	} else {
+		// if expected variables but cannot be extracted from the string.
+		if len(vars) > 0 {
+			return &apis.FieldError{
+				Message: fmt.Sprintf("Expected to find variables but none found in %q", value),
+				// Empty path is required to make the `ViaField`, … work
+				Paths: []string{""},
+			}
+		}
```

Comment:
Adding this `else` will not work because different values can be passed to `ValidateVariableP` function other than just param names i.e. image name, step name. 

```yaml
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  generateName: validation-
spec:
  taskSpec:
    params:
      - name: foo
        default: "bar1"
    results:
      - name: echo-output
        description: "successful echo"
    steps:
      - name: echo-params
        image: bash
        script: |
          set -e
          echo $(params.foo) | tee $(results.bad-echo.path)
```

Try this example, when you apply, you'll see 
```
Error from server (BadRequest): error when creating "validation.yaml": admission webhook "validation.webhook.pipeline.tekton.dev" denied the request: validation failed: Expected to find variables but none found in "": spec.taskSpec.steps[0].workingDir
Expected to find variables but none found in "bash": spec.taskSpec.steps[0].image
Expected to find variables but none found in "echo-params": spec.taskSpec.steps[0].name
Expected to find variables but none found in "set -e\necho $(params.foo) | tee $(results.bad-echo.path)\n": spec.taskSpec.steps[0].script
```